### PR TITLE
Add project for gcs audits logs

### PR DESCRIPTION
**Pull request by ameukam**
- Dedicated GCP project for PII
- BigQuery dataset for PII analysis
- Bucket for GCR container registries audit logs  

Signed-off-by: Arnaud Meukam <ameukam@gmail.com>

### CODE REVIEW

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+  service = each.key
+}
+
+
+// BigQuery dataset for audit logs
+resource "google_bigquery_dataset" "audit-logs-gcs" {
+  dataset_id                  = local.bucket_name
+  friendly_name               = local.bucket_name
+  delete_contents_on_destroy  = true
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:58]:
```suggestion
  delete_contents_on_destroy  = false
```
I mean maybe we should treat this as an ephemeral cache of data that can be arbitrarily destroyed at any time? But I personally err on the side of "data is sacred"

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+resource "google_bigquery_dataset_iam_member" "bq-sinks-binding" {
+  project    = google_project.project.project_id
+  dataset_id = google_bigquery_dataset.audit-logs-gcs.id
+  role       = "roles/bigquery.dataEditor"
+  member     = google_logging_organization_sink.gcs-logs-org-sink.writer_identity
+
+}
+
+# A bucket to store logs in audit logs for GCS
+resource "google_storage_bucket" "audit-logs-gcs" {
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:75]: Another resource I might protect from deletion on destroy

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+// Allow ready-only access to k8s-infra-gcs-access-logs@kubernetes.io
+resource "google_storage_bucket_iam_member" "artificats-gcs-logs" {
+  bucket = google_storage_bucket.audit-logs-gcs.name
+  role   = "roles/storage.objectViewer"
+  member = "group:k8s-infra-gcs-access-logs@kubernetes.io"
+}
+
+// Create a sink for the organization
+resource "google_logging_organization_sink" "gcs-logs-org-sink" {
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:107]: Not familiar with this so no opinion

**Review by spiffxp — CHANGES_REQUESTED** [review]

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf (hunk 3 above)
Comment (ameukam) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:107]: I was looking at [Aggregated sinks](https://cloud.google.com/logging/docs/export/aggregated_sinks) but I realized I don't need this.

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  billing_account = data.google_billing_account.account.id
+}
+
+resource "google_project_service" "project" {
+  project = google_project.project.id
+
+  for_each = toset([
+    "bigquery.googleapis.com",
+    "bigqueryreservation.googleapis.com",
+    "bigquerytransfer.googleapis.com",
```
Comment (ameukam) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:48]: @spiffxp I will use this service to transfer the data from `k8s-infra-ii-sandbox:k8s_artifacts_gcslogs_appspot` to `k8s-infra-public-pii:k8s-infra-k8s_infra_artifacts_gcslogs`.

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  project_id  = "k8s-infra-public-pii"
+  bucket_name = "k8s-infra-artifacts-gcslogs"
```
Comment (ameukam) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:21]: Naming is really hard. I tempted to switch to `k8s-infra-gcr-audit-logs`. Not sure if it's for all the GCR registries or just scoped to `k8s-artifacts-prod`. 
@spiffxp WDYT ?

Comment (spiffxp) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:21]: tl;dr I'm happy with `k8s-infra-artifacts-gcslogs`.  

>Not sure if it's for all the GCR registries

GCR doesn't generate any access logs, and there are non-GCR buckets in k8s-artifacts-prod, e.g. https://github.com/kubernetes/k8s.io/blob/main/audit/projects/k8s-artifacts-prod/buckets/k8s-artifacts-kind/logging.json.  

> or just scoped to k8s-artifacts-prod.

Thinking with a mindset of bucket-level IAM vs. object-level ACLs, I wouldn't anticipate putting anything more than access logs related to artifacts here.  We might find we want to do something similar for `k8s-staging` buckets, so I'm fine not constraining to `artifacts-prod` in the name, but staying more general with `artifacts`

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  service = each.key
+}
+
+// Ensure k8s-infra-gcp-auditors@kubernetes.io can administer all the Biquery
+// resources except the datasets and the datasets tables
+resource "google_project_iam_member" "auditors-bq-resources-admins" {
+  project = google_project.project.id
+  role    = "roles/bigquery.resourceAdmin"
+  member  = "group:k8s-infra-gcp-auditors@kubernetes.io"
+}
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:61]: I might be wrong, but I'm not sure I want a traditionally read-only group to be able to create things like capacityCommitments or reservationAssignments.

They should already have all the necessary permissions to export metadata (like which datasets exist, and who's able to access them) via the `audit.viewer` role assigned at the organization level.  They should not be able to view the data within the datasets (same mindset with buckets: they are not granted read access to all objects in all buckets, but they are granted list access)
```suggestion
```

ref: https://cloud.google.com/bigquery/docs/access-control

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  project                     = google_project.project.project_id
+}
+
+
+// Ensure k8s-infra-gcp-auditors@kubernetes.io has admin access to the dataset
+resource "google_bigquery_dataset_iam_member" "auditors-bq-dataset-editors" {
+  dataset_id = local.dataset-id
+  role       = "roles/bigquery.dataEditor"
+  member     = "group:k8s-infra-gcp-auditors@kubernetes.io"
+}
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:79]: Definitely don't want to give auditors access to PII
```suggestion
```

Alternatively, I'd be ok with `k8s-infra-gcs-access-logs@kubernetes.io` getting this level of access

File: infra/gcp/clusters/projects/k8s-infra-public-pii/provider.tf [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/*
+This file defines:
+- Required provider versions
+- Storage backend details
+*/
+
+terraform {
+
+  backend "gcs" {
+    bucket = "k8s-infra-tf-public-ii"
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/provider.tf:10]: Or maybe I'm misunderstanding the intent here.  I assumed this was about PII, not about the `ii` folks having a dedicated project in public
```suggestion
    bucket = "k8s-infra-tf-public-pii"
```

File: infra/gcp/ensure-main-project.sh [hunk 9]
```diff
@@ -75,6 +75,7 @@ readonly TERRAFORM_STATE_BUCKET_ENTRIES=(
     k8s-infra-tf-aws:k8s-infra-aws-admins@kubernetes.io
     k8s-infra-tf-prow-clusters:k8s-infra-prow-oncall@kubernetes.io
     k8s-infra-tf-public-clusters:"${CLUSTER_ADMINS_GROUP}"
+    k8s-infra-tf-public-ii:"${CLUSTER_ADMINS_GROUP}"
```
Comment (spiffxp) [inline infra/gcp/ensure-main-project.sh:78]: Same as above
```suggestion
    k8s-infra-tf-public-pii:"${CLUSTER_ADMINS_GROUP}"
```

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  // NOTE: Prevent the bucket from being deleted
+  lifecycle {
+    prevent_destroy = false
+  }
+}
+
+// Grant write access to group-cloud-storage-analytics@google.com
+resource "google_storage_bucket_acl" "analytics_write" {
```
Comment (spiffxp) [inline infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf:106]: Can we do this with IAM instead of ACLs?

**Review by spiffxp — CHANGES_REQUESTED** [review]

File: infra/gcp/clusters/projects/k8s-infra-public-pii/provider.tf (hunk 8 above)
Comment (ameukam) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/provider.tf]: Fixed

File: infra/gcp/ensure-main-project.sh (hunk 9 above)
Comment (ameukam) [reply infra/gcp/ensure-main-project.sh]: Fixed

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf (hunk 10 above)
Comment (ameukam) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf]: Done.

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf (hunk 6 above)
Comment (ameukam) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf]: Removed.

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf (hunk 7 above)
Comment (ameukam) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf]: Removed.

File: infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf (hunk 1 above)
Comment (ameukam) [reply infra/gcp/clusters/projects/k8s-infra-public-pii/main.tf]: Fixed

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by ameukam** [comment]
Running `./ensure-main-project.sh`

**Comment by spiffxp** [comment]
`Ensuring 'gs://k8s-infra-tf-public-pii' exists as private with owners 'k8s-infra-cluster-admins@kubernetes.io'`

Didn't catch this during review... this is too broad a group.  I will open a followup to move this to org-admins once this has been deployed

**Comment by spiffxp** [comment]
@ameukam I was going to run `terraform apply` but I'm going to hold off if you're taking on deployment

**Comment by ameukam** [comment]
Got during the deployment : 

```bash
Error: Error setting IAM policy for storage bucket "b/k8s-infra-artifacts-gcslogs": googleapi: Error 409: The metadata for object "null" was edited during the operation. Please try again., conflict

  on main.tf line 114, in resource "google_storage_bucket_iam_policy" "analytics_legacybucketwriter_policy":
 114: resource "google_storage_bucket_iam_policy" "analytics_legacybucketwriter_policy" {
```

I think the bindings with `cloud-storage-analytics@google.com` as member are in conflict.

**Pull request merged**